mock/py/mockbuild/buildroot.py

```diff
@@ -497,15 +498,39 @@ def _setup_devices(self):
             kver = os.uname()[2]
             self.root_log.debug("kernel version == %s", kver)
             for i in devFiles:
+                src_path = "/" + i[2]
+                chroot_path = self.make_chroot_path(i[2])
+
+                if util.cmpKernelVer(kver, '2.6.18') >= 0 and src_path == '/dev/ptmx':
+                    continue
+
                 # create node, but only if it exist on host too
                 # except for loop devices, which only show up on the host after they are first used
-                if os.path.exists("/" + i[2]) or "loop" in i[2]:
-                    os.mknod(self.make_chroot_path(i[2]), i[0], i[1])
+                if os.path.exists(src_path) or "loop" in src_path:
+                    try:
+                        os.mknod(chroot_path, i[0], i[1])
+                    except OSError as e:
+                        # If mknod gives us a permission error, fall back to a different
+                        # strategy of using a bind mount from root to host. This won't
+                        # work for the loop devices, so just skip them in this case.
+                        if e.errno == errno.EPERM:
```

Comment:
Hmm. This is done even on bare metals. Not just for containers. Previously if the mknod failed on bare metal, it failed. Now we bind mount host device. I am not saying it is bad (not sure actually), but it is definitely change from past.

Reply:
It's fairly difficult to detect "on bare metal" reliably without encoding assumptions about the container system (docker, podman, lxd, systemd-nspawn...), so I've avoided "in a container" conditionalization anywhere in the patchset.

And conditionalizing based on !USE_NSPAWN as Pavel suggests makes the assumption that 
systemp-nspawn will never ever work nested in a container - which I don't think is necessarily going to be true forever.

IMO, we can make the code simpler and more future proof by just trying the fallback in all cases - maybe it will help if we get an EPERM even on bare metal because of some other reason, e..g, selinux seccomp.

Making error reporting more confusing is always a concern, but it's not the code is expecting mknod to fail and producing a nice informative error message - it's just dying with a traceback in that case.